src/aws-cpp-sdk-core/include/aws/core/utils/stream/AwsChunkedStream.h

```diff
@@ -0,0 +1,93 @@
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/crypto/Hash.h>
+
+namespace Aws {
+namespace Utils {
+namespace Stream {
+
+static const size_t AWS_DATA_BUFFER_SIZE = 65536;
+
+template <size_t DataBufferSize = AWS_DATA_BUFFER_SIZE>
+class AwsChunkedStream {
+ public:
+  AwsChunkedStream(Http::HttpRequest *request, const std::shared_ptr<Aws::IOStream> &stream)
+      : m_chunkingStream{Aws::MakeShared<StringStream>("AwsChunkedStream")}, m_request(request), m_stream(stream) {
+    assert(m_stream != nullptr);
+    if (m_stream == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "stream is null");
+    }
+    assert(request != nullptr);
+    if (request == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "request is null");
+    }
+  }
+
+  size_t BufferedRead(char *dst, size_t amountToRead) {
+    assert(dst != nullptr);
+    if (dst == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "dst is null");
+    }
+
+    // the chunk has ended and cannot be read from
+    if (m_chunkEnd) {
+      return 0;
+    }
+
+    // If we've read all of the underlying stream write the checksum trailing header
+    // the set that the chunked stream is over.
+    if (m_stream->eof() && !m_stream->bad() && (m_chunkingStream->eof() || m_chunkingStream->peek() == EOF)) {
+      Aws::StringStream chunkedTrailerStream;
+      chunkedTrailerStream << "0\r\n";
+      if (m_request->GetRequestHash().second != nullptr) {
+        chunkedTrailerStream << "x-amz-checksum-" << m_request->GetRequestHash().first << ":"
+                             << HashingUtils::Base64Encode(m_request->GetRequestHash().second->GetHash().GetResult()) << "\r\n";
+      }
+      chunkedTrailerStream << "\r\n";
+      const auto chunkedTrailer = chunkedTrailerStream.str();
+      auto trailerSize = chunkedTrailer.size();
+      assert(amountToRead >= trailerSize);
+      memcpy(dst, chunkedTrailer.c_str(), trailerSize);
+      m_chunkEnd = true;
+      return trailerSize;
+    }
+
+    // Try to read in a 64K chunk, if we cant we know the stream is over
+    size_t bytesRead = 0;
+    while (m_stream->good() && bytesRead < DataBufferSize) {
+      m_stream->read(&m_data[bytesRead], DataBufferSize - bytesRead);
+      bytesRead += m_stream->gcount();
+    }
+
+    // update the trailing checksum to be sent only if we read data and buffered.
+    if (bytesRead > 0 && m_request->GetRequestHash().second != nullptr) {
+      m_request->GetRequestHash().second->Update(reinterpret_cast<unsigned char *>(m_data.GetUnderlyingData()), bytesRead);
+    }
+
+    // Buffer chunked encoding from data if there was data read to the buffer, otherwise leave it alone/
+    if (bytesRead > 0 && m_chunkingStream != nullptr && !m_chunkingStream->bad()) {
+      *m_chunkingStream << Aws::Utils::StringUtils::ToHexString(bytesRead) << "\r\n";
+      std::copy(m_data.GetUnderlyingData(), m_data.GetUnderlyingData() + bytesRead, std::ostream_iterator<char>(*m_chunkingStream));
```

Comment:
this will reduce copy by avoiding copying it to an intermediate std::ostream_iterator
 m_chunkingStream->write(m_data.GetUnderlyingData(), bytesRead);